File: tools/parity/src/main/kotlin/io/github/glandais/parity/PipelineDump.kt
package io.github.glandais.parity

import io.github.glandais.engine.Enhancer
import io.github.glandais.engine.gpx.GpxParser
import io.github.glandais.engine.gpx.firstTrackAsPath
import io.github.glandais.engine.path.ElevationStep
import io.github.glandais.engine.path.Path
import io.github.glandais.engine.path.PathSimplifier
import io.github.glandais.engine.path.PointPerDistance
import io.github.glandais.engine.path.PointPerSecond
import io.github.glandais.engine.physics.MaxSpeedComputer
import io.github.glandais.engine.physics.VirtualizeService
import java.io.File

/**
 * Kotlin side of the end-to-end parity cascade.
 *
 * Re-implements [Enhancer.enhanceCourse] step by step (same calls, same order) so the
 * [Path] can be dumped *between* stages — a divergence at stage N is only diagnosable if
 * stage N-1 is known to match.
 *
 * Keep in lockstep with :
 * - `engine/src/commonMain/.../Enhancer.kt` (the behaviour being mirrored)
 * - `tools/parity/ts/pipelineDump.ts` (the TS counterpart)
 *
 * `fixElevation` is intentionally not wired here : the deterministic cascade runs with no
 * network, and elevation parity is measured separately (see `tools/parity/README.md`).
 */
private class StageWriter(
    private val outDir: File,
) {
    private var index = 0
    var timeOrigin: Double = 0.0

    fun dump(
        stage: String,
        path: Path,
    ) {
        DumpFormat.writeStage(outDir, index, stage, path, timeOrigin)
        System.err.println("  [kt] %02d-%s: %d pts".format(index, stage, path.size))
        index++
    }
}

fun main(argv: Array<String>) {
    val args = argv.toList()

    fun arg(name: String): String? {
        val i = args.indexOf("--$name")
        return if (i >= 0 && i + 1 < args.size) args[i + 1] else null
    }

    val gpxFile = arg("gpx") ?: error("usage: --gpx <file.gpx> --out <dir> [--simplify]")
    val outDir = arg("out") ?: error("usage: --gpx <file.gpx> --out <dir> [--simplify]")
    val doSimplify = args.contains("--simplify")

    val writer = StageWriter(File(outDir))

    // ---- parse ---------------------------------------------------------------
    var path = GpxParser.parse(File(gpxFile).readText()).firstTrackAsPath()

    // Normalise the time axis to t0 = 0, symmetrically with the TS runner. Harness
    // decision, not a pipeline one : it keeps the absolute epoch (~1.7e12 ms) out of the
    // comparison, where it would swamp millisecond-level differences at float64
    // resolution. The offset is recorded in every header so nothing is lost.
    if (path.size > 0) {
        writer.timeOrigin = path.time(0)
        for (i in 0 until path.size) {
            path.setTime(i, path.time(i) - writer.timeOrigin)
        }
        path.computeDerivedData()
    }
    writer.dump("parsed", path)

    // ---- pipeline, mirroring Enhancer.enhanceCourse ---------------------------
    val course = Enhancer.getDefaultCourse(path)

    path = PointPerDistance.compute(path, minDistanceM = -1.0, maxDistanceM = 30.0)
    writer.dump("ppd-30", path)

    path = PointPerDistance.compute(path, minDistanceM = 1.0, maxDistanceM = 2.0)
    writer.dump("ppd-2", path)

    path = ElevationStep.smoothElevation(path)
    writer.dump("smooth", path)

    var working = course.copy(course = course.course.copy(path = path))
    MaxSpeedComputer.computeMaxSpeeds(working.course) // mutates `path` in place
    writer.dump("maxspeed", path)

    path = VirtualizeService.virtualizeTrack(working)
    working = working.copy(course = working.course.copy(path = path))
    writer.dump("virtualize", path)

    path = PointPerSecond.computeOnePointPerSecond(path)
    writer.dump("pointpersecond", path)

    if (doSimplify) {
        path = PathSimplifier.simplify(path, 10.0, 3.0)
        writer.dump("simplify", path)
    }
}

File: tools/parity/ts/pipelineDump.ts
/**
 * TypeScript side of the end-to-end parity cascade.
 *
 * Re-implements `Enhancer.enhanceCourse` step by step (same calls, same order, same
 * arguments) so the `Path` can be dumped *between* stages. Diverging at stage N is only
 * diagnosable if stage N-1 is known to match.
 *
 * Keep in lockstep with:
 *   - virtual-cyclist/src/enhancer/Enhancer.ts  (the behaviour being mirrored)
 *   - tools/parity/src/main/kotlin/.../PipelineDump.kt  (the Kotlin counterpart)
 *
 * Usage (cwd MUST be virtual-cyclist/, so tsx picks up its tsconfig `paths`):
 *   npx tsx <this> --gpx <file.gpx> --out <dir> [--simplify] [--fix-elevation]
 */
import './bootstrap';

import { readFileSync } from 'node:fs';
import { Enhancer } from '@/enhancer/';
import { GPXParser } from '@/gpx/';
import { MaxSpeedComputer, VirtualizeService } from '@/physics/';
import { DouglasPeucker, PointPerDistance, PointPerSecond } from '@/processing/';
import { CoursePhysics } from '@/types/course/';
import { Path, POINT_FIELDS } from '@/types/path/';
import { Elevation } from '@/elevation/';
import { DumpHeader, writeStage } from './dumpFormat';

const FIELD_NAMES = [
    'latitude', 'longitude', 'distance', 'dx',
    'time', 'elapsed', 'dt',
    'bearing',
    'elevation',
    'grade',
    'radius',
    'aeroCoef',
    'windBearing', 'windAlpha',
    'pAero', 'pGravity', 'pRollingResistance', 'pWheelBearings',
    'pInputPower', 'pCyclistProvidedOptimalPower',
    'pCyclistProvidedOptimalPowerWithHarmonics', 'pCyclistPowerNeeded',
    'pCyclistProvidedMuscular', 'pCyclistProvidedWheel',
    'pComputedTotalPower', 'pComputedWheelPower', 'pComputedPower',
    'speed', 'speedMax', 'speedMaxIncline', 'virtSpeedCurrent',
    'temperature', 'windSpeed', 'windDirection',
    'heartRate', 'cadence',
];

function arg(name: string): string | undefined {
    const i = process.argv.indexOf(`--${name}`);
    return i >= 0 ? process.argv[i + 1] : undefined;
}
function flag(name: string): boolean {
    return process.argv.includes(`--${name}`);
}

/**
 * Simulation start timestamp used by `clock-pinned` mode.
 *
 * `VirtualizeService.ts` defaults its simulation clock to `new Date().getTime()`, so the TS
 * time axis starts around 1.785e12 ms, where float64 quantises to 2.44e-4 ms — about 2.1e6
 * times coarser than Kotlin's axis, which starts at 0. That alone shifts `dt`, `speed` and
 * every power field by ~1e-7 relative from the very first point, and gives `PointPerSecond`
 * a different sub-second phase to snap to. Without pinning, the measurement only ever
 * reproduces the TS clock defect (divergence 1 in docs/parity.md).
 *
 * Since `virtual-cyclist` 1.3.0 (`8532ba0`, *feat: allow custom start time for track
 * virtualization*) this is a **supported parameter** — `virtualizeTrack(course, startTime)`,
 * also reachable as `EnhanceOptions.startTime`. The harness used to monkey-patch `globalThis
 * .Date` instead; that is gone. Nothing ambient is overridden any more, and the pinned mode
 * now exercises the shipped API exactly as a caller would.
 */
const PINNED_START_MS = 0;

let stageIndex = 0;
let timeOrigin = 0;

function dump(outDir: string, stage: string, path: Path): void {
    const size = path.getPointCount();
    let durationMs = 0;
    if (size >= 2) {
        durationMs = path.getField(size - 1, 4 /* TIME */) - path.getField(0, 4 /* TIME */);
    }
    const header: DumpHeader = {
        stage,
        index: stageIndex++,
        size,
        fieldCount: POINT_FIELDS.length,
        fields: FIELD_NAMES,
        totalDistance: path.getTotalDistance(),
        durationMs,
        elevationGain: path.getTotalElevationGain(),
        elevationLoss: path.getTotalElevationLoss(),
        minElevation: size === 0 ? 0 : path.getMinElevation(),
        maxElevation: size === 0 ? 0 : path.getMaxElevation(),
        timeOrigin,
    };
    writeStage(outDir, header, (i, f) => path.getField(i, POINT_FIELDS[f]));
    process.stderr.write(`  [ts] ${String(header.index).padStart(2, '0')}-${stage}: ${size} pts\n`);
}

async function main(): Promise<void> {
    const gpxFile = arg('gpx');
    const outDir = arg('out');
    if (!gpxFile || !outDir) {
        throw new Error('usage: --gpx <file.gpx> --out <dir> [--simplify] [--fix-elevation]');
    }
    const doSimplify = flag('simplify');
    const doFixElevation = flag('fix-elevation');
    const zeroClock = flag('zero-clock');

    // ---- parse ----------------------------------------------------------------
    const parsed = GPXParser.parse(readFileSync(gpxFile, 'utf-8'));
    if (parsed.tracks.length === 0) throw new Error(`no track in ${gpxFile}`);
    let path = parsed.tracks[0];

    // Normalise the time axis to t0 = 0, symmetrically with the Kotlin runner. This is a
    // harness decision, not a pipeline one: it keeps the absolute epoch (~1.7e12 ms) out of
    // the comparison, where it would swamp millisecond-level differences in float64
    // resolution. The offset is recorded in every header so nothing is lost.
    if (path.getPointCount() > 0) {
        timeOrigin = path.getField(0, 4 /* TIME */);
        for (let i = 0; i < path.getPointCount(); i++) {
            path.setField(i, 4 /* TIME */, path.getField(i, 4) - timeOrigin);
        }
        path.computeDerivedData();
    }
    dump(outDir, 'parsed', path);

    // ---- pipeline, mirroring Enhancer.enhanceCourse ----------------------------
    const course: CoursePhysics = Enhancer.getDefaultCourse(path);

    path = PointPerDistance.compute(path, -1.0, 30.0, Enhancer.INPUT_GPX_FIELDS);
    dump(outDir, 'ppd-30', path);

    if (doFixElevation) {
        path = await Elevation.fixElevation(path);
        dump(outDir, 'fixelevation', path);
    }

    path = PointPerDistance.compute(path, 1.0, 2.0, Enhancer.INPUT_GPX_FIELDS);
    dump(outDir, 'ppd-2', path);

    path = await Elevation.smoothElevation(path);
    dump(outDir, 'smooth', path);

    const courseWithPath: CoursePhysics = { ...course, path };
    MaxSpeedComputer.computeMaxSpeeds(courseWithPath); // mutates `path` in place
    dump(outDir, 'maxspeed', path);

    // `undefined` lets the library apply its own `new Date()` default, which is exactly what
    // `as-is` mode is meant to measure.
    path = VirtualizeService.virtualizeTrack(courseWithPath, zeroClock ? PINNED_START_MS : undefined);
    dump(outDir, 'virtualize', path);

    path = PointPerSecond.computeOnePointPerSecond(path);
    dump(outDir, 'pointpersecond', path);

    if (doSimplify) {
        path = DouglasPeucker.simplify(path, 10, 3);
        dump(outDir, 'simplify', path);
    }
}

main().catch(err => {
    process.stderr.write(`${err instanceof Error ? err.stack : String(err)}\n`);
    process.exit(1);
});
